### ISSUE TITLE
hyphen (U+2010) vs. en dash (U+2013) vs. em dash (U+2014)

### ISSUE THREAD

**Issue by chrissimpkins** [source-foundry/legible#5]
Issue report: 

- https://github.com/chrissimpkins/Hack/issues/178

**Comment by iandoug** [comment]
en dash and em dash seem to be problematic.
I've added them to this comparison, which is a part of a curated list of my previous comparison of confusing characters, having deleted all the fonts which fail in one of more places for being "confusing".
Putting en and em next to each other sometimes gets a merge, and in some fonts the em is clearly NOT the the same width as other characters, so I guess it is going over the char box boundary.
![spectacle v29809](https://user-images.githubusercontent.com/12673886/27984537-e3fa9d86-63d8-11e7-9ab7-4ed4320abf59.png)

**Comment by chrissimpkins** [comment]
This is helpful Ian.  Looks like the hyphen and en are similar in size across most of the typefaces in that image.  In a few cases, en slightly longer but shorter than em.  In Eco, looks like the opposite is the case.  Only one that I see where en shorter than hyphen.  

em looks like it is nearly full box width in all.   

Do you know of a scenario where these would be adjacent in source code, markup, or other commonly used bodies of text for monospaced fonts?  

Looks like we have quite a few more ambiguous combinations to add as IR.  Any particularly concerning ones that you would like to add?

**Comment by iandoug** [comment]
Just read that Oxygen Mono is lacking en and em dash so I guess the ones above are substitutes.

I was actually surprised that it is an issue because I thought that using those dashes would be restricted to people doing "typographic" work eg in a word processor or desktop publishing rather than typing code in a text editor. I have never come across code that requires en or em dashes. IIRC original poster was using non-English and maybe situation is different for him.

I have added copyright and @ to compare. Have been staring at the keyboard to see what else COULD be confusing. I know when I was looking for fonts to label a keyboard there were some where pipe could be confused with upper-I or lower-l but that doesn't seem to be such a problem with the Monos. Also as previously pointed out some fonts manage to get either ({ or [{ looking very similar.
Some fonts (like Fira Code) manage to get [] to make a box.
Some serif fonts have an issue with ec at low point sizes, probably more an issue with italics.

Off topic but don't know where else to mention, but Fira Code didn't make the cut (mainly because of the tilde), and I think Ahamono does a good job of differentiation, apart from the hyphen-endash. But everyone is going to struggle with those. Similarly with the numerous single-quote symbols that Unicode saw fit to define. 

I suppose running en or em at an angle is going to look funny. Don't know if you have KCharSelect (or similar) but load something like Everson Mono (because of glyph coverage) and do searches for "hyphen" or "dash" and see what pops up, as well as the "See also" lists. There is much room for confusion here.

**Comment by chrissimpkins** [comment]
> copyright and @

good idea!

> I suppose running en or em at an angle is going to look funny

This is the kind of idea that I hope the conversations here generate.  Not necessarily true and is something worth investigation if we feel that this is an important problem.  It will not look typical, but in some circumstances that may not be a bad thing from a legilibity standpoint.  An example of this is the broken upright pipe glyph that David van Gemeren designed in the Hack italic set to address the issue of the italic pipe (previously angled straight line) being confused for a forward slash.  We found that this was a rare situation in the body of programming language source code, but was an issue in comments associated with the languages where many syntax highlighters will use italics. Some do not like the aesthetics of it (and for a more fierce dislike of aesthetics, see the comments in issue report from our Cyrillic users about the change that I attempted to the Cyrillic c with the intent to distinguish this from a Latin c), but one of our primary goals has been to try to maintain the Bitstream Vera Mono - DejaVu Sans Mono - Hack lineage style and improve upon the legibility of the glyphs in source.

**Comment by iandoug** [comment]
BTW "Amber" was my test for "font not found". So don't waste your time trying to find Amber Mono :-)

**Comment by iandoug** [comment]
Okay I see what you mean with italic slash.
However UK keyboards have the broken bar character too (as well as the Not sign)
So is the Hack solution not potentially confusing?

[https://en.wikipedia.org/wiki/Vertical_bar#Solid_vertical_bar_vs_broken_bar](url)

**Comment by iandoug** [comment]
[https://www.cs.sfu.ca/~ggbaker/reference/characters/](url)

also the "The Trouble With EM 'n EN" linked to from there.

**Comment by chrissimpkins** [comment]
Re: https://github.com/source-foundry/legible/issues/5#issuecomment-313890110

Interesting.  I wasn't aware that there was a broken bar character defined by Unicode.  @burodepeper , we may need to revisit the italic pipe?  It looks like we are using a Unicode defined shape.

**Comment by chrissimpkins** [comment]
Re: https://www.cs.sfu.ca/~ggbaker/reference/characters/

Good resource!  Any suggestions about how to include these here?  Maybe a directory that includes a library of images/URL/etc?

**Comment by burodepeper** [comment]
@chrissimpkins I thought you knew about the broken pipe.

@iandoug The broken pipe on a UK keyboard, do you mean that it actually prints a broken pipe when you press it, or is it merely printed on your keyboard as a broken pipe, but instead prints a regular pipe? I have ANSI keycaps, and it uses a broken pipe for a regular pipe to distinguish itself from the backslash that's on the same key. I'll ask some friends in the UK about their keyboards.

**Comment by iandoug** [comment]
I have a few URLs etc to put in a file. 
But first I need to figure out how to get my fork of this repo back in synch with yours.
At the moment my limited experience of git says I must do this:
1. clone it to my local machine
2. do a sync from there to your repo
3. clone from my local machine to my copy on github

Which seems a rather convoluted way of doing things (as opposed to "click the magic SYNCHRONISE button" which I can't find ...

So yeah will then create a doc/whatever with links/resources and also upload the program.

**Comment by iandoug** [comment]
@burodepeper 
Judging from that doc above (and the Wikipedia page) I think the answer is "it depends".
On the UK boards it seems to be on AltGr, so not a "commonly used" character. It may be historical.
I suspect the ASCII art guys may make use of that character.
I vaguely remember from DOS / OS/2 days that they used to display the pipe as a broken pipe, and my memory banks also say I had a keyboard with both pipe and broken pipe printed, and possibly on the same key. This may date back to the 84 key keyboards rather than the now-standard 104 versions.

Went surfabout and this popped up, people complaining about recent UK MS keyboards:
[http://techhead.co/misleading-pipe-symbol-location-on-microsoft-keyboard/](url)
where image and character produced are opposite.

So I say "it depends" because it probably depends on what your keyboard driver does with that keyboard scancode. And hopefully the Linux guys did a better job than MS. :-)

**Comment by chrissimpkins** [comment]
How is broken pipe used?

**Comment by chrissimpkins** [comment]
@iandoug push to your fork remote and click create pull request button on your repository 

https://help.github.com/articles/creating-a-pull-request/

**Comment by iandoug** [comment]
Re broken pipe: It is used, but now in our universe.... 
Last time I worked in EBCDIC was back in the 1990s...

"I am reading from a mainframe file using power exchange. I have the copybook also available. However the fields are "broken pipe" dilimited ¦ and NOT the regular | symbol. How can I read this file. I am sure there must be a way to do it. Any help will be greatly appreciated"
[https://network.informatica.com/thread/18174](url)

But I'm sure there are people on PCs who also use it, just as a defence against clueless users messing with their stuff, the same way that they would use the "NOT" sign as a separator, eg in csv files (where I use a tilde as the resulting files are much less hassle than having to worry about putting quotes around strings with commas).

**Comment by chrissimpkins** [comment]
I see.  Thanks for the explanation.  Thoughts David? Seems low priority to change from where we are given that.   

My suggestion was going to be to slant the break since it was an italic glyph.

**Comment by chrissimpkins** [comment]
If anyone feels compelled to modify it let's break it out into a separate IR

**Comment by burodepeper** [comment]
> The broken bar has hardly any practical application and does not appear to have any clearly identified uses distinct from the vertical bar.

I think there's no reason to look into it at the moment. It is simply not used in programming (which is focused around ASCII). Besides that, in a context, it is clear what the character means, as far as I'm concerned. We could indeed slant it, but then I would argue to break the regular pipe as well, to make it consistent. Anyway, low priority.

**Comment by iandoug** [comment]
FWIW, quick comparison. There may be font substitution taking place with broken pipe.
![spectacle j11385](https://user-images.githubusercontent.com/12673886/27996917-3a944c7c-64ed-11e7-8807-172ad4984cc4.png)

**Comment by burodepeper** [comment]
@iandoug Could you render an italic version that also includes a backslash? That is the character that become too similar to a regular pipe and sparked the discussion.

**Comment by iandoug** [comment]
Attached.
It's done via the little PHP program I uploaded here:
[https://github.com/source-foundry/legible/issues/7](url)
Get the latest version at the bottom, you can adjust out (characters, fonts, colours, etc).
![spectacle tj1831](https://user-images.githubusercontent.com/12673886/27997520-c7ace2ca-64f9-11e7-9ece-80b57a77487d.png)

**Comment by chrissimpkins** [comment]
That render looks fantastic!

**Comment by chrissimpkins** [comment]
Looks like italics are in bold-italic?

**Comment by iandoug** [comment]
Yes, there are options for normal, italic, oblique, bold, and bolditalic.
You want normal italic?
Attached.
![spectacle j27480](https://user-images.githubusercontent.com/12673886/27997600-097f9e84-64fc-11e7-9e36-27cc5849a717.png)

**Comment by iandoug** [comment]
FWIW I think I have all four variants of Hack installed, but for the others, I only have the Regular version installed, so Firefox is doing the bold/italic bits....
